### DISCUSSION TITLE
Error: No QueryClient set, use QueryClientProvider to set one

### DISCUSSION THREAD

**Discussion by HashiW** [TanStack/query#1461] [Q&A]
I am trying to use a simple useQuery Hook however it always produce the above error. I believe this is due to the new updates and changes in v3. How should I be using useQuery in V3 ?

![image](https://user-images.githubusercontent.com/41080448/102446181-3268c400-3ffb-11eb-9421-e03b9ac4d951.png)

**Desktop (please complete the following information):**
 - Windows
 - Opera
 - Version: [e.g. 22]

**Comment by TkDodo** [comment]
Have you read the migration guide?
https://react-query.tanstack.com/guides/migrating-to-react-query-3

As the error suggests, you need to add a QueryClientProvider.

**Reply by Chijioke91** [reply]
@HashiW Can you check where you are placing your dev-tool, It is meant to be inside your QueryClientProvider....That was my error

For example mine was previously

```
<Fragment>
   <QueryClientProvider client={queryClient}>
      <MyApp/>
    </QueryClientProvider>
<ReactQueryDevtools initialIsOpen={false} />
</Fragment>
```

But I changed it to

```
<Fragment>
   <QueryClientProvider client={queryClient}>
      <MyApp/>
     <ReactQueryDevtools initialIsOpen={false} />
   </QueryClientProvider>
</Fragment>
```

Then everything worked...Hope this works for you too

**Reply by juancarlosqr** [reply]
Same error... misplacing the `ReactQueryDevtools`  😕 Thanks

**Reply by nairo-mudumane** [reply]
I had the same error, thank you for helping

**Reply by dannyoka** [reply]
Had the same error! Thanks!

**Comment by faran4engg** [comment]
You can mock `useQuery` as below:

`jest.mock('react-query', () => ({
  useQuery: () => ({
    isLoading: false,
    error: {},
    data: [],
  }),
}));`

**Comment by Karnak19** [comment]
I have the same error aswell. I'm really confused, it seems nothing is misplaced. 🤕 
![Capture d’écran 2021-01-28 à 11 18 36](https://user-images.githubusercontent.com/7700494/106123756-a221b100-615a-11eb-9633-e080e672b8b3.png)


```jsx
// index.jsx

const queryClient = new QueryClient();

ReactDOM.render(
  <StrictMode>
    <QueryClientProvider queryClient={queryClient}>
      <Player />
      {process.env.NODE_ENV === "development" && <ReactQueryDevtools />}
    </QueryClientProvider>
  </StrictMode>,
  document.getElementById("root")
);
```

```js
// api.js

export const getSongs = () => axios.get(`${API_URL}/songs`);
```

```jsx
// player.jsx

import { getSongs } from "./api";

function Player() {
// .......
  const { isLoading, error, data, isFetching } = useQuery("songs", getSongs);
// .......
```

**Desktop (please complete the following information):**
- macOS Big Sur 11.1 (Apple M1)
- Firefox 85.0b9 / Chrome 88.0
- react-query 3.6.0
- react 17.0.1
- react-scripts 4.0.1


Edit: After deleting and re-installing node_modules, it works ✅  🤷

**Comment by carloskiki** [comment]
I am getting the same error, I don't understand everything looks in place:
```
import React from "react";
import { QueryClient, QueryClientProvider } from "react-query";

import Routes from "./src/Routes";

const queryClient = new QueryClient();

export default function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <Routes />
    </QueryClientProvider>
  );
}
```
Desktop:
- MacOS 11.2 (Apple M1)
- react-query 3.13.4
- react 17.0.2
- react-scripts 4.0.3
- react-native/expo SDK 40.0.1

**Reply by Kayyow** [reply]
Same for me with:

```
import React from 'react'
import Router from './Router'
import { QueryClient, QueryClientProvider } from 'react-query'
import { ReactQueryDevtools } from 'react-query/devtools'

const queryClient = new QueryClient()

const App = () => (
  <QueryClientProvider client={queryClient}>
    <Router />
    <ReactQueryDevtools initialIsOpen={false} />
  </QueryClientProvider>
)

export default App
```

- macOS 11.2.3 (Intel Core i7)
- react-query 3.13.6
- react 17.0.2
- vite 2.1.5

**Reply by TkDodo** [reply]
vite has problems with the devtools. have a look here: https://github.com/tannerlinsley/react-query/pull/2113

**Reply by Kayyow** [reply]
Thank you for your quick answer

**Comment by icyJoseph** [comment]
I think for a bunch of people this might happen if you are following Tan's course. In which case, you should probably install v2 instead, or try to learn and update as you go, until he releases the v3 course.

**Comment by ManikantaMaddipati** [comment]
I am facing the same issue. : 
https://stackoverflow.com/questions/67582043/error-no-queryclient-set-use-queryclientprovider-to-set-one

**Reply by TkDodo** [reply]
and I think you found the solution already ;)

**Comment by revskill10** [comment]
Got this issue with nextjs 12

**Reply by icyJoseph** [reply]
Which setup do you have? I am using next 12 too and have no issues what so ever.

**Comment by bennik88** [comment]
I'm having the same issue with ` "next": "12.2.4"`, ` "firebase": "^9.9.2"`, `"@react-query-firebase/firestore": "^1.0.0-dev.7"`, `"@tanstack/react-query": "^4.1.0"` And can't get rid of it. For sure added the provider like in the docs.

**Reply by TkDodo** [reply]
react-query-firebase/firestore needs react-query v3 as peerDependency:
- https://github.com/invertase/react-query-firebase/blob/0acc863d8c5a89974fca64307ddddbba173e72d9/packages/firestore/package.json#L44

so likely you wind up with two versions of react-query.

**Reply by asifsaho** [reply]
I am having this issue with my next js setup. I have the provider in `_app.tsx` but still complaining.

```js
import createCache from "@emotion/cache";
import { CacheProvider } from "@emotion/react";
import CssBaseline from "@mui/material/CssBaseline";
import { StyledEngineProvider, ThemeProvider } from "@mui/material/styles";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import { SessionProvider } from "next-auth/react";
import { AppProps } from "next/app";
import Head from "next/head";
import { useState } from "react";
import { RecoilRoot } from "recoil";
import { theme } from "theme";
import { staticStr } from "utils/staticStrings";
import "./global.css";

export default function MyApp(props: AppProps) {
  const {
    Component,
    pageProps: { session, ...pageProps },
  } = props;

  const queryClient = new QueryClient();

  return (
    <SessionProvider session={session}>
      <StyledEngineProvider injectFirst>
        <CacheProvider value={cache}>
          <Head>
            <title>{staticStr.title}</title>
            <meta
              name="viewport"
              content="minimum-scale=1, initial-scale=1, width=device-width"
            />
          </Head>
          <StyledEngineProvider injectFirst>
            <ThemeProvider theme={theme}>
              <CssBaseline />
              <RecoilRoot>
                <QueryClientProvider client={queryClient}>
                  <>
                    <Component {...pageProps} />
                    <ReactQueryDevtools />
                  </>
                </QueryClientProvider>
              </RecoilRoot>
            </ThemeProvider>
          </StyledEngineProvider>
        </CacheProvider>
      </StyledEngineProvider>
    </SessionProvider>
  );
}
```

**Reply by chrift** [reply]
I am also having this error using a set-up similar to the above, except the `queryClient` instantiation is outside of the app function

Next v12
react-query 3.34.19
@tanstack/react-query-devtools 4.3.5

Have tried bumping react-query to the latest version, and also switching to @tanstack/react-query 4.3.4, still get the same error

**Reply by asifsaho** [reply]
My problem is solved, I needed to remove the old version. Somehow two versions were there and created an issue.

**Comment by chrift** [comment]
Following on from my reply: https://github.com/TanStack/query/discussions/1461#discussioncomment-3646727

I switched back to 
react-query 3.34.19
and instead of installing the devtools separately, I am importing them like:
`import { ReactQueryDevtools } from 'react-query/devtools'`

Now works.

**Reply by moyongshi** [reply]
react-query is old,
the new is 
 "@tanstack/react-query": "^4.3.4",
 "@tanstack/react-query-devtools": "^4.3.5",

**Reply by MoAnsir** [reply]
This fixed my problem too. I had 
`import { ReactQueryDevtools } from "@tanstack/react-query-devtools";`
And was breaking my app

**Reply by gabrielperman** [reply]
Same here. My guess is that `import { ReactQueryDevtools } from "@tanstack/react-query-devtools";` is only compatible with v4 as of now and v3 should use the bundled `import { ReactQueryDevtools } from 'react-query/devtools'`

@moyongshi v4 is still in beta. v3 is still the stable version.

**Reply by TkDodo** [reply]
> @moyongshi v4 is still in beta. v3 is still the stable version.

where do you have that information from please? We need to clear up this confusion. v4 was released half a year ago:

- https://tanstack.com/blog/announcing-tanstack-query-v4
- https://tanstack.com/query/v4/docs/react/overview
- https://www.npmjs.com/package/@tanstack/react-query

Latest version is `4.20.4` already! The v3 docs redirect to the v4 docs 😅

**Comment by Peroconino** [comment]
I had the same error but with in my monorepo setup, so basically the project i was working with had tanstack/react-query installed locally in the project and the same project also imports another workspace project with tanstack/react-query installed there too, so what have i done to get it fixed was using that workspace to export the reactQueryProvider and use on my main project. Thats kinda weird that happens but it worked.

PS : im using pnpm as package manager

**Reply by TkDodo** [reply]
I had the same issue with pnpm yesterday. FYI, it's because pnpm resolves dependencies that have peerDependencies (like react-query) "differently" and you might wind up with multiple copies of the same version of react-query in node_modules, which explains the context issue.

For me, the issue was solved by _also_ adding react and react-dom as peerDependencies to the lib:

- https://github.com/pnpm/pnpm/issues/5351#issuecomment-1273610927

**Reply by Peroconino** [reply]
I have something in that conditions that might be related to this problem, basically what happens is: if i do not import the types of useMutation or useQuery from react-query it fails to reffer and gives me a error:
![Screenshot_1](https://user-images.githubusercontent.com/50136672/195086009-4b7d09f5-7a26-407d-aaad-3eace0f25579.png)

But if i import UseQueryResult from react-query it stops.

**Reply by iamrakhmatov** [reply]
I have exact same monorepo description as `Leandro` but I am getting this issue because of `generated graphql file naming`

**ex:**  
  If file name `products.graphql` it generates `products.generated.ts` and there is no error
  but If file name is `getProducts.graphql` it generates `getProducts.generated.ts` and there is an error inside the file same as following ->

<img width="1347" alt="Screen Shot 2023-01-03 at 11 41 17 PM" src="https://user-images.githubusercontent.com/61634173/210422133-8865306e-50ea-4911-ba30-5f56095cb364.png">

Using:
- Turborepo 
- @tanstack/react-query: ^4.19.1  _(on both `main` app and `shared-graphql` package)_

FYI: 
- `main` app uses `vite`
- `pnpm` as a package manager

**Reply by iamrakhmatov** [reply]
Sorry!!!
**Correction:** 
This wasn't related to `file naming` it just gives this error anyway despite the name of the file

**Reply by iamrakhmatov** [reply]
but in my case I am getting two errors ->

<img width="975" alt="Screen Shot 2023-01-04 at 12 24 34 AM" src="https://user-images.githubusercontent.com/61634173/210427014-08c4888b-4e89-4396-8468-2feddfdefdad.png">

@TkDodo is there a way to solve this without tweaking peerDependencies?

**Comment by jacobgoh101** [comment]
On my end, it works after I update both `@tanstack/react-query` & `@tanstack/react-query-devtools` to the latest version.

**Comment by 00155837364765867** [comment]
Context providers can be a real pain in JS, especially when building / exporting a module.
If you're seeing this error in your published node_modules (i.e. you're writing a module for use in another App), this _might_ be the issue (there are so many ways for this error to show up!):
As most are no doubt aware, the QueryClientProvider needs to be 'wrapped around' the file that refs useQuery - i.e. it cannot be in the same file.
For a regular App, this isn't an issue, as the wrapper is the App component. For a module, you likely don't have an 'App' - rather just a bunch of exports.
So...create a separate wrapper file and export that. I also found adding contextSharing={true} is needed...
Like this:

`const MyModuleWrapper = (props) => {

    return (
        <QueryClientProvider client={queryClient} contextSharing={true}>
            <MyModule/>
        </QueryClientProvider>
    );
}`

Hopefully, with some luck this might help someone. It may not....this issue can rear its head in many different ways!

**Comment by azmaat1** [comment]
Thanks

**Comment by jose-codegourmet** [comment]
I was able to fix this by double checking on the imports 

**_app.tsx**
```
import {
  DehydratedState,
  Hydrate,
  QueryClient,
  QueryClientProvider,
} from "@tanstack/react-query";
```

**posts.tsx**

`import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";`


hope this helps

**Comment by mauriau** [comment]
Hello, I have the same issue. It's works with queries, but bug with mutation :/
I found my mistake, it was the import.
I used `import {useMutation} from "react-query";
`
but it's
import `{ useMutation } from '@tanstack/react-query';`
